File: crates/ferrogate-providers/src/anthropic.rs
// Token4AI Cloud Attribution
// Developed by the commercial cloud service company represented by https://token4ai.cloud.
// Author: jamesduan (X: https://x.com/JamesDuanL)
// Created: 2026-06-11
// description: Token4AI Cloud, FerroGate AI Gateway, Rust API Gateway, agent-native AI traffic infrastructure.

use serde_json::{json, Value};

use ferrogate_core::{ToolCall, ToolDef, ToolResult};

use crate::{
    canonical::CanonicalAiRequest, AdapterError, ChatCompletionPlan, ProviderAdapter,
    ProviderConfig, ProviderErrorResponse, ProviderHeader, ProviderHttpRequest, ProviderUsage,
    ResponsesPlan, SecretValue,
};

#[derive(Debug, Default, Clone, Copy)]
pub struct AnthropicAdapter;

impl ProviderAdapter for AnthropicAdapter {
    fn kind(&self) -> &'static str {
        "anthropic"
    }

    fn prepare_chat_completions(
        &self,
        provider: ProviderConfig,
        request: ChatCompletionPlan,
    ) -> Result<ProviderHttpRequest, AdapterError> {
        validate_kind(&provider.kind)?;
        let body = ensure_object_body(request.body)?;
        let messages = body.get("messages").cloned().unwrap_or_else(|| json!([]));
        let max_tokens = body
            .get("max_tokens")
            .cloned()
            .unwrap_or_else(|| json!(1024));

        let mut anthropic_body = json!({
            "model": request.provider_model,
            "messages": messages,
            "max_tokens": max_tokens,
            "stream": request.stream,
        });
        if let Some(system) = body.get("system") {
            anthropic_body["system"] = system.clone();
        }

        Ok(ProviderHttpRequest {
            provider: provider.name,
            endpoint: messages_endpoint(&provider.base_url),
            body: anthropic_body,
            stream: request.stream,
            headers: anthropic_headers(provider.api_key),
        })
    }

    fn prepare_responses(
        &self,
        provider: ProviderConfig,
        request: ResponsesPlan,
    ) -> Result<ProviderHttpRequest, AdapterError> {
        validate_kind(&provider.kind)?;
        let body = CanonicalAiRequest::from_responses_body(request.body)?.into_anthropic_body();
        let messages = body.get("messages").cloned().unwrap_or_else(|| json!([]));
        let max_tokens = body
            .get("max_tokens")
            .cloned()
            .unwrap_or_else(|| json!(1024));

        let mut anthropic_body = json!({
            "model": request.provider_model,
            "messages": messages,
            "max_tokens": max_tokens,
            "stream": request.stream,
        });
        if let Some(system) = body.get("system") {
            anthropic_body["system"] = system.clone();
        }
        if let Some(tools) = body.get("tools") {
            anthropic_body["tools"] = tools.clone();
        }
        if let Some(tool_choice) = body.get("tool_choice") {
            anthropic_body["tool_choice"] = tool_choice.clone();
        }

        Ok(ProviderHttpRequest {
            provider: provider.name.clone(),
            endpoint: messages_endpoint(&provider.base_url),
            body: anthropic_body,
            stream: request.stream,
            headers: anthropic_headers(provider.api_key),
        })
    }

    fn normalize_error_response(
        &self,
        status: u16,
        content_type: &str,
        body: &[u8],
        request_id: &str,
    ) -> ProviderErrorResponse {
        let parsed = serde_json::from_slice::<Value>(body).ok();
        let provider_error = parsed.as_ref().and_then(|value| value.get("error"));
        let message = provider_error
            .and_then(|error| error.get("message"))
            .and_then(Value::as_str)
            .map(ToOwned::to_owned)
            .or_else(|| fallback_error_message(parsed.as_ref(), body))
            .unwrap_or_else(|| format!("provider returned HTTP {status}"));
        let provider_type = provider_error
            .and_then(|error| error.get("type"))
            .and_then(Value::as_str)
            .unwrap_or("provider_error");

        ProviderErrorResponse {
            status,
            body: json!({
                "error": {
                    "message": message,
                    "type": "provider_error",
                    "provider_type": provider_type,
                    "code": provider_type,
                    "provider_status": status,
                    "provider_content_type": content_type,
                    "request_id": request_id,
                }
            }),
        }
    }

    fn extract_usage(&self, body: &[u8]) -> Option<ProviderUsage> {
        let value = serde_json::from_slice::<Value>(body).ok()?;
        let usage = value.get("usage").or_else(|| {
            value
                .get("message")
                .and_then(|message| message.get("usage"))
        })?;
        let prompt_tokens = usage.get("input_tokens").and_then(Value::as_u64);
        let completion_tokens = usage.get("output_tokens").and_then(Value::as_u64);
        let total_tokens = prompt_tokens
            .zip(completion_tokens)
            .map(|(left, right)| left + right);
        let extracted = ProviderUsage {
            prompt_tokens,
            completion_tokens,
            total_tokens,
        };
        (extracted.prompt_tokens.is_some()
            || extracted.completion_tokens.is_some()
            || extracted.total_tokens.is_some())
        .then_some(extracted)
    }

    fn inject_tools(&self, body: Value, tools: &[ToolDef]) -> Result<Value, AdapterError> {
        let mut body = ensure_object_body(body)?;
        if tools.is_empty() {
            return Ok(body);
        }
        body["tools"] = Value::Array(
            tools
                .iter()
                .map(|tool| {
                    let mut value = json!({
                        "name": tool.name,
                        "input_schema": tool.input_schema,
                    });
                    if let Some(description) = &tool.description {
                        value["description"] = Value::String(description.clone());
                    }
                    value
                })
                .collect(),
        );
        Ok(body)
    }

    fn extract_tool_calls(&self, body: &[u8]) -> Result<Vec<ToolCall>, AdapterError> {
        let value = serde_json::from_slice::<Value>(body).map_err(|error| {
            AdapterError::InvalidRequest {
                message: format!("provider response body must be JSON: {error}"),
            }
        })?;
        Ok(value
            .get("content")
            .and_then(Value::as_array)
            .into_iter()
            .flatten()
            .filter(|block| {
                block
                    .get("type")
                    .and_then(Value::as_str)
                    .is_some_and(|kind| kind == "tool_use")
            })
            .filter_map(anthropic_tool_use_to_canonical)
            .collect())
    }

    fn append_tool_results(
        &self,
        body: Value,
        results: &[ToolResult],
    ) -> Result<Value, AdapterError> {
        let mut body = ensure_object_body(body)?;
        if results.is_empty() {
            return Ok(body);
        }
        if !body.get("messages").is_some_and(Value::is_array) {
            body["messages"] = json!([]);
        }
        let messages = body["messages"]
            .as_array_mut()
            .expect("messages initialized as array");
        messages.push(json!({
            "role": "user",
            "content": results.iter().map(|result| {
                json!({
                    "type": "tool_result",
                    "tool_use_id": result.tool_call_id,
                    "content": tool_result_content_to_string(&result.content),
                    "is_error": result.is_error,
                })
            }).collect::<Vec<_>>(),
        }));
        Ok(body)
    }
}

fn anthropic_headers(api_key: Option<String>) -> Vec<ProviderHeader> {
    let mut headers = vec![
        ProviderHeader {
            name: "content-type".into(),
            value: SecretValue::new("application/json"),
        },
        ProviderHeader {
            name: "anthropic-version".into(),
            value: SecretValue::new("2023-06-01"),
        },
    ];
    if let Some(api_key) = api_key.filter(|value| !value.trim().is_empty()) {
        headers.push(ProviderHeader {
            name: "x-api-key".into(),
            value: SecretValue::new(api_key),
        });
    }
    headers
}

fn validate_kind(kind: &str) -> Result<(), AdapterError> {
    match kind {
        "anthropic" => Ok(()),
        other => Err(AdapterError::UnsupportedProviderKind {
            kind: other.to_string(),
        }),
    }
}

fn ensure_object_body(body: Value) -> Result<Value, AdapterError> {
    if body.is_object() {
        Ok(body)
    } else {
        Err(AdapterError::InvalidRequest {
            message: "chat completion request body must be a JSON object".into(),
        })
    }
}

fn messages_endpoint(base_url: &str) -> String {
    format!("{}/messages", base_url.trim_end_matches('/'))
}

fn fallback_error_message(parsed: Option<&Value>, body: &[u8]) -> Option<String> {
    parsed
        .and_then(|value| value.as_str().map(ToOwned::to_owned))
        .or_else(|| {
            let text = std::str::from_utf8(body).ok()?.trim();
            (!text.is_empty()).then(|| text.chars().take(512).collect())
        })
}

fn anthropic_tool_use_to_canonical(value: &Value) -> Option<ToolCall> {
    Some(ToolCall {
        id: value.get("id")?.as_str()?.to_string(),
        name: value.get("name")?.as_str()?.to_string(),
        arguments: value.get("input").cloned().unwrap_or(Value::Null),
    })
}

fn tool_result_content_to_string(value: &Value) -> String {
    value
        .as_str()
        .map(ToOwned::to_owned)
        .unwrap_or_else(|| value.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn provider(api_key: Option<&str>) -> ProviderConfig {
        ProviderConfig {
            name: "anthropic".into(),
            kind: "anthropic".into(),
            base_url: "https://api.anthropic.example/v1/".into(),
            api_key: api_key.map(str::to_string),
            openrouter_http_referer: None,
            openrouter_x_title: None,
            aws_credentials: None,
            gcp_credentials: None,
        }
    }

    #[test]
    fn converts_openai_chat_plan_to_anthropic_messages_request() {
        let adapter = AnthropicAdapter;
        let prepared = adapter
            .prepare_chat_completions(
                provider(Some("provider-secret")),
                ChatCompletionPlan {
                    logical_model: "claude-chat".into(),
                    provider_model: "claude-3-5-sonnet-latest".into(),
                    stream: true,
                    body: json!({
                        "model": "claude-chat",
                        "messages": [{"role": "user", "content": "hello"}],
                        "max_tokens": 256,
                        "system": "be concise"
                    }),
                },
            )
            .unwrap();

        assert_eq!(
            prepared.endpoint,
            "https://api.anthropic.example/v1/messages"
        );
        assert_eq!(prepared.body["model"], "claude-3-5-sonnet-latest");
        assert_eq!(prepared.body["messages"][0]["content"], "hello");
        assert_eq!(prepared.body["max_tokens"], 256);
        assert_eq!(prepared.body["system"], "be concise");
        assert_eq!(prepared.body["stream"], true);
        assert!(prepared
            .headers
            .iter()
            .any(|header| header.name == "x-api-key"
                && header.value.expose_secret() == "provider-secret"));
        assert!(!format!("{prepared:?}").contains("provider-secret"));
    }

    #[test]
    fn rejects_non_object_chat_body() {
        let adapter = AnthropicAdapter;
        let error = adapter
            .prepare_chat_completions(
                provider(None),
                ChatCompletionPlan {
                    logical_model: "claude-chat".into(),
                    provider_model: "claude".into(),
                    stream: false,
                    body: json!("bad"),
                },
            )
            .unwrap_err();

        assert_eq!(
            error,
            AdapterError::InvalidRequest {
                message: "chat completion request body must be a JSON object".into()
            }
        );
    }

    #[test]
    fn converts_responses_plan_to_anthropic_messages_request() {
        let adapter = AnthropicAdapter;
        let prepared = adapter
            .prepare_responses(
                provider(Some("provider-secret")),
                ResponsesPlan {
                    logical_model: "claude-chat".into(),
                    provider_model: "claude-3-5-sonnet-latest".into(),
                    stream: false,
                    body: json!({
                        "model": "claude-chat",
                        "instructions": "be concise",
                        "input": "hello",
                        "max_output_tokens": 256
                    }),
                },
            )
            .unwrap();

        assert_eq!(
            prepared.endpoint,
            "https://api.anthropic.example/v1/messages"
        );
        assert_eq!(prepared.body["model"], "claude-3-5-sonnet-latest");
        assert_eq!(prepared.body["messages"][0]["role"], "user");
        assert_eq!(prepared.body["messages"][0]["content"], "hello");
        assert_eq!(prepared.body["system"], "be concise");
        assert_eq!(prepared.body["max_tokens"], 256);
        assert_eq!(prepared.body["stream"], false);
    }

    #[test]
    fn converts_responses_plan_with_tool_choice_and_image_inputs_to_anthropic_messages_request() {
        let adapter = AnthropicAdapter;
        let prepared = adapter
            .prepare_responses(
                provider(None),
                ResponsesPlan {
                    logical_model: "claude-chat".into(),
                    provider_model: "claude-3-5-sonnet-latest".into(),
                    stream: false,
                    body: json!({
                        "tools": [{
                            "type": "function",
                            "function": {
                                "name": "lookup_weather",
                                "parameters": {
                                    "type": "object",
                                    "properties": {"city": {"type": "string"}}
                                }
                            }
                        }],
                        "tool_choice": {"type": "tool", "name": "lookup_weather"},
                        "input": [{
                            "role": "user",
                            "content": [
                                {"type": "input_text", "text": "look"},
                                {"type": "input_image", "image_url": "https://example.com/a.png"}
                            ]
                        }]
                    }),
                },
            )
            .unwrap();

        assert_eq!(prepared.body["tools"][0]["name"], "lookup_weather");
        assert_eq!(prepared.body["tool_choice"]["type"], "tool");
        assert_eq!(prepared.body["messages"][0]["content"][0]["type"], "text");
        assert_eq!(prepared.body["messages"][0]["content"][1]["type"], "image");
        assert_eq!(
            prepared.body["messages"][0]["content"][1]["source"]["type"],
            "url"
        );
    }

    #[test]
    fn normalizes_anthropic_error_response() {
        let adapter = AnthropicAdapter;
        let normalized = adapter.normalize_error_response(
            400,
            "application/json",
            br#"{"type":"error","error":{"type":"invalid_request_error","message":"bad request"}}"#,
            "fg-test",
        );

        assert_eq!(normalized.status, 400);
        assert_eq!(normalized.body["error"]["message"], "bad request");
        assert_eq!(
            normalized.body["error"]["provider_type"],
            "invalid_request_error"
        );
        assert_eq!(normalized.body["error"]["code"], "invalid_request_error");
    }

    #[test]
    fn extracts_anthropic_usage_metadata() {
        let adapter = AnthropicAdapter;
        let usage = adapter
            .extract_usage(br#"{"usage":{"input_tokens":13,"output_tokens":8}}"#)
            .unwrap();

        assert_eq!(usage.prompt_tokens, Some(13));
        assert_eq!(usage.completion_tokens, Some(8));
        assert_eq!(usage.total_tokens, Some(21));
    }

    #[test]
    fn injects_anthropic_tools_and_round_trips_tool_calls() {
        let adapter = AnthropicAdapter;
        let body = adapter
            .inject_tools(
                json!({"model": "claude", "messages": []}),
                &[ToolDef {
                    name: "lookup_weather".into(),
                    description: Some("Lookup weather".into()),
                    input_schema: json!({
                        "type": "object",
                        "properties": {"city": {"type": "string"}}
                    }),
                }],
            )
            .unwrap();

        assert_eq!(body["tools"][0]["name"], "lookup_weather");
        assert_eq!(
            body["tools"][0]["input_schema"]["properties"]["city"]["type"],
            "string"
        );

        let calls = adapter
            .extract_tool_calls(
                br#"{"stop_reason":"tool_use","content":[{"type":"text","text":"checking"},{"type":"tool_use","id":"toolu_1","name":"lookup_weather","input":{"city":"Shanghai"}}]}"#,
            )
            .unwrap();
        assert_eq!(calls.len(), 1);
        assert_eq!(calls[0].id, "toolu_1");
        assert_eq!(calls[0].name, "lookup_weather");
        assert_eq!(calls[0].arguments["city"], "Shanghai");
    }

    #[test]
    fn appends_anthropic_tool_results_as_content_blocks() {
        let adapter = AnthropicAdapter;
        let body = adapter
            .append_tool_results(
                json!({"messages": [{"role": "user", "content": "weather?"}]}),
                &[ToolResult {
                    tool_call_id: "toolu_1".into(),
                    content: json!({"temp_c": 21}),
                    is_error: false,
                }],
            )
            .unwrap();

        assert_eq!(body["messages"][1]["role"], "user");
        assert_eq!(body["messages"][1]["content"][0]["type"], "tool_result");
        assert_eq!(body["messages"][1]["content"][0]["tool_use_id"], "toolu_1");
        assert_eq!(
            body["messages"][1]["content"][0]["content"],
            r#"{"temp_c":21}"#
        );
    }
}

File: packages/providers/src/anthropic.ts
/**
 * Anthropic provider adapter — port of `anthropic.rs`.
 *
 * Dispatches chat completions and Responses plans as Anthropic `/v1/messages`
 * requests (the latter via {@link CanonicalAiRequest}), and normalizes
 * errors/usage/tool-calls back to the canonical gateway shape.
 */
import type { ToolCall, ToolDef, ToolResult } from "@ferrogate/core";

import { messageToChatCompletion } from "./anthropic_messages.js";
import { applyPromptCacheToAnthropic, promptCacheFromBody } from "./caching.js";
import { CanonicalAiRequest } from "./canonical.js";
import { asStr, asU64, getField, isObject, ownBody, parseJson } from "./json.js";
import type { Json, JsonObject } from "./json.js";
import { fallbackErrorMessage, hasAnyUsage } from "./openai.js";
import { applyStructuredOutputToAnthropic, structuredOutputFromChatBody } from "./structured.js";
import { AdapterError, BaseProviderAdapter, SecretValue } from "./types.js";
import type {
  ChatCompletionPlan,
  ProviderConfig,
  ProviderErrorResponse,
  ProviderHeader,
  ProviderHttpRequest,
  ProviderUsage,
  ResponsesPlan,
} from "./types.js";

export class AnthropicAdapter extends BaseProviderAdapter {
  override kind(): string {
    return "anthropic";
  }

  override prepareChatCompletions(
    provider: ProviderConfig,
    request: ChatCompletionPlan,
  ): ProviderHttpRequest {
    validateKind(provider.kind);
    const body = ensureObjectBody(request.body);
    const messages = getField(body, "messages") ?? [];
    const maxTokens = getField(body, "max_tokens") ?? 1024;

    const draft: JsonObject = {
      model: request.providerModel,
      messages,
      max_tokens: maxTokens,
      stream: request.stream,
    };
    const system = getField(body, "system");
    if (system !== undefined) draft["system"] = system;
    // `messages` and `system` in the draft are still the CALLER's objects,
    // shared with every other candidate on the failover ladder. `ownBody` takes
    // a private deep copy, so the breakpoints and coercion tools placed below
    // decorate THIS request and no other (issue #690). It is also the only way
    // to obtain the argument the `apply*` helpers accept, so the copy cannot be
    // skipped by the next adapter that needs to rewrite something.
    const anthropicBody = ownBody(draft);
    // `response_format` has no Anthropic equivalent, so it used to be dropped
    // here while surviving to an OpenAI upstream — the failover shape change of
    // issue #674. It is now coerced into a forced tool call, or refused.
    const structured = structuredOutputFromChatBody(body);
    if (structured !== undefined) {
      applyStructuredOutputToAnthropic(anthropicBody, structured, provider.kind);
    }
    // Prompt caching (issue #690). Applied AFTER the structured-output coercion
    // so a coercion tool is inside the cached prefix rather than appended after
    // the breakpoint, where it would invalidate the cache on every request.
    const promptCache = promptCacheFromBody(body);
    if (promptCache !== undefined) {
      applyPromptCacheToAnthropic(anthropicBody, promptCache, provider.kind);
    }

    return {
      provider: provider.name,
      endpoint: messagesEndpoint(provider.baseUrl),
      body: anthropicBody,
      stream: request.stream,
      headers: anthropicHeaders(provider.apiKey),
    };
  }

  override prepareResponses(provider: ProviderConfig, request: ResponsesPlan): ProviderHttpRequest {
    validateKind(provider.kind);
    const body = CanonicalAiRequest.fromResponsesBody(request.body).intoAnthropicBody();
    const messages = getField(body, "messages") ?? [];
    const maxTokens = getField(body, "max_tokens") ?? 1024;

    const draft: JsonObject = {
      model: request.providerModel,
      messages,
      max_tokens: maxTokens,
      stream: request.stream,
    };
    const system = getField(body, "system");
    if (system !== undefined) draft["system"] = system;
    const tools = getField(body, "tools");
    if (tools !== undefined) draft["tools"] = tools;
    const toolChoice = getField(body, "tool_choice");
    if (toolChoice !== undefined) draft["tool_choice"] = toolChoice;
    // Owned for the same reason as the chat path: `intoAnthropicBody` copies the
    // Responses body shallowly, so `system`/`tools` here can still be subtrees
    // the caller — and every other candidate — holds.
    const anthropicBody = ownBody(draft);

    return {
      provider: provider.name,
      endpoint: messagesEndpoint(provider.baseUrl),
      body: anthropicBody,
      stream: request.stream,
      headers: anthropicHeaders(provider.apiKey),
    };
  }

  override translateChatCompletionResponse(body: Uint8Array, model: string): Json | null {
    const message = parseJson(body);
    return isObject(message) ? messageToChatCompletion(message, model) : null;
  }

  override normalizeErrorResponse(
    status: number,
    contentType: string,
    body: Uint8Array,
    requestId: string,
  ): ProviderErrorResponse {
    const parsed = parseJson(body);
    const providerError = getField(parsed, "error");
    const message =
      asStr(getField(providerError, "message")) ??
      fallbackErrorMessage(parsed, body) ??
      `provider returned HTTP ${status}`;
    const providerType = asStr(getField(providerError, "type")) ?? "provider_error";

    return {
      status,
      body: {
        error: {
          message,
          type: "provider_error",
          provider_type: providerType,
          code: providerType,
          provider_status: status,
          provider_content_type: contentType,
          request_id: requestId,
        },
      },
    };
  }

  override extractUsage(body: Uint8Array): ProviderUsage | undefined {
    const value = parseJson(body);
    if (value === undefined) return undefined;
    const usage = getField(value, "usage") ?? getField(getField(value, "message"), "usage");
    if (usage === undefined) return undefined;
    const promptTokens = asU64(getField(usage, "input_tokens"));
    const completionTokens = asU64(getField(usage, "output_tokens"));
    const totalTokens =
      promptTokens !== undefined && completionTokens !== undefined
        ? promptTokens + completionTokens
        : undefined;
    const extracted: ProviderUsage = { promptTokens, completionTokens, totalTokens };
    return hasAnyUsage(extracted) ? extracted : undefined;
  }

  override injectTools(body: Json, tools: readonly ToolDef[]): Json {
    const object = ensureObjectBody(body);
    if (tools.length === 0) return object;
    object["tools"] = tools.map((tool) => {
      const value: JsonObject = { name: tool.name, input_schema: tool.input_schema as Json };
      if (tool.description !== undefined) value["description"] = tool.description;
      return value;
    });
    return object;
  }

  override extractToolCalls(body: Uint8Array): ToolCall[] {
    const value = parseJson(body);
    if (value === undefined) {
      throw AdapterError.invalidRequest("provider response body must be JSON");
    }
    const content = getField(value, "content");
    if (!Array.isArray(content)) return [];
    const calls: ToolCall[] = [];
    for (const block of content) {
      if (asStr(getField(block, "type")) === "tool_use") {
        const canonical = anthropicToolUseToCanonical(block);
        if (canonical) calls.push(canonical);
      }
    }
    return calls;
  }

  override appendToolResults(body: Json, results: readonly ToolResult[]): Json {
    const object = ensureObjectBody(body);
    if (results.length === 0) return object;
    if (!Array.isArray(object["messages"])) object["messages"] = [];
    const messages = object["messages"] as Json[];
    messages.push({
      role: "user",
      content: results.map((result) => ({
        type: "tool_result",
        tool_use_id: result.tool_call_id,
        content: toolResultContentToString(result.content as Json),
        is_error: result.is_error,
      })),
    });
    return object;
  }
}

function anthropicHeaders(apiKey: string | undefined): ProviderHeader[] {
  const headers: ProviderHeader[] = [
    { name: "content-type", value: new SecretValue("application/json") },
    { name: "anthropic-version", value: new SecretValue("2023-06-01") },
  ];
  if (apiKey !== undefined && apiKey.trim().length > 0) {
    headers.push({ name: "x-api-key", value: new SecretValue(apiKey) });
  }
  return headers;
}

function validateKind(kind: string): void {
  if (kind !== "anthropic") throw AdapterError.unsupportedProviderKind(kind);
}

function ensureObjectBody(body: Json): JsonObject {
  if (isObject(body)) return body;
  throw AdapterError.invalidRequest("chat completion request body must be a JSON object");
}

const trimEndSlashes = (value: string): string => value.replace(/\/+$/, "");
const messagesEndpoint = (baseUrl: string): string => `${trimEndSlashes(baseUrl)}/messages`;

function anthropicToolUseToCanonical(value: Json): ToolCall | undefined {
  const id = asStr(getField(value, "id"));
  const name = asStr(getField(value, "name"));
  if (id === undefined || name === undefined) return undefined;
  return { id, name, arguments: getField(value, "input") ?? null };
}

const toolResultContentToString = (value: Json): string =>
  typeof value === "string" ? value : JSON.stringify(value);
